pkg/api/universal/conversation_stream.go

```diff
@@ -0,0 +1,312 @@
+}
+
+// StreamingPipelineImpl handles the streaming conversation pipeline
+type StreamingPipelineImpl struct {
+	middleware []StreamingMiddleware
+	logger     logger.Logger
+}
+
+// StreamingMiddleware is a simple interface for processing streaming chunks
+type StreamingMiddleware interface {
+	ProcessChunk(chunk []byte) []byte
+	Flush() []byte
+}
+
+// NewStreamingPipelineImpl creates a new streaming pipeline
+func NewStreamingPipelineImpl(logger logger.Logger) *StreamingPipelineImpl {
+	return &StreamingPipelineImpl{
+		middleware: make([]StreamingMiddleware, 0),
+		logger:     logger,
+	}
+}
+
+// AddMiddleware adds a middleware to the pipeline
+func (p *StreamingPipelineImpl) AddMiddleware(middleware StreamingMiddleware) {
+	if middleware == nil {
+		return
+	}
+
+	p.middleware = append(p.middleware, middleware)
+}
+
+// ProcessStream handles the streaming pipeline from LangChain Go to gRPC stream
+func (p *StreamingPipelineImpl) ProcessStream(
+	ctx context.Context,
+	stream runtimev1pb.Dapr_ConverseStreamAlpha1Server,
+	component conversation.Conversation,
+	request *conversation.ConversationRequest,
+) (*runtimev1pb.ConversationUsage, error) {
+	// Try to get the underlying LangChain Go model for streaming support
+	if streamer, ok := component.(conversation.StreamingConversation); ok {
+		return p.processRealStreaming(ctx, stream, streamer, request)
+	}
+
+	// Fallback: simulate streaming with existing Converse method
+	return p.processSimulatedStreaming(ctx, stream, component, request)
+}
+
+// processRealStreaming handles true LangChain Go streaming
+func (p *StreamingPipelineImpl) processRealStreaming(
+	ctx context.Context,
+	stream runtimev1pb.Dapr_ConverseStreamAlpha1Server,
+	streamer conversation.StreamingConversation,
+	request *conversation.ConversationRequest,
+) (*runtimev1pb.ConversationUsage, error) {
+	var finalResp *conversation.ConversationResponse
+	var streamErr error
+
+	// Create the streaming function that LangChain Go will call for each chunk
+	streamFunc := func(chunkCtx context.Context, chunk []byte) error {
+		// Process chunk through middleware chain
+		processedChunk := p.processChunkThroughMiddleware(chunk)
+
+		// Send processed chunk immediately (if any ready)
+		if len(processedChunk) > 0 {
+			if err := p.sendChunk(stream, processedChunk); err != nil {
+				p.logger.Errorf("Failed to send streaming chunk: %v", err)
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	// Call the component's streaming method
+	finalResp, streamErr = streamer.ConverseStream(ctx, request, streamFunc)
+
+	// Flush any remaining content in middleware buffers
+	if remaining := p.flushMiddleware(); len(remaining) > 0 {
+		if sendErr := p.sendChunk(stream, remaining); sendErr != nil {
+			p.logger.Errorf("Failed to send final streaming chunk: %v", sendErr)
+			// Return the component error if it exists, otherwise the send error
+			if streamErr != nil {
+				return nil, streamErr
+			}
+			return nil, sendErr
+		}
+	}
```

Comment:
It seems somewhat clumsy to need to be flushing here, _after_ we have returned from `ConverseStream`.